ee/server/service/users.go

```diff
@@ -61,3 +101,55 @@ func (svc *Service) GetSSOUser(ctx context.Context, auth fleet.Auth) (*fleet.Use
 	}
 	return user, nil
 }
+
+// rolesChanged checks whether there was any change between the old and new roles.
+//
+// rolesChanged assumes that there cannot be multiple role entries for the same team.
+func rolesChanged(oldGlobal *string, oldTeams []fleet.UserTeam, newGlobal *string, newTeams []fleet.UserTeam) bool {
+	if (newGlobal != nil && (oldGlobal == nil || *oldGlobal != *newGlobal)) || (newGlobal == nil && oldGlobal != nil) {
+		return true
+	}
+	if len(oldTeams) != len(newTeams) {
+		return true
+	}
+	oldTeamsMap := make(map[uint]fleet.UserTeam)
+	for _, oldTeam := range oldTeams {
+		oldTeamsMap[oldTeam.Team.ID] = oldTeam
+	}
+	for _, newTeam := range newTeams {
+		oldTeam, ok := oldTeamsMap[newTeam.Team.ID]
+		if !ok {
+			return true
+		}
+		if oldTeam.Role != newTeam.Role {
+			return true
+		}
+	}
+	return false
+}
+
+// userRolesFromSSOAttributes loads the global or team roles from custom SSO attributes.
+//
+// The returned `globalRole` and `teamRoles` are ready to be assigned to
+// `fleet.User` struct fields `GlobalRole` and `Teams` respectively.
+//
+// If the custom attributes are not found, then the default global observer ("observer", nil, nil) is returned.
+func (svc *Service) userRolesFromSSOAttributes(ctx context.Context, auth fleet.Auth) (globalRole *string, teamsRoles []fleet.UserTeam, err error) {
+	ssoRolesInfo, err := fleet.RolesFromSSOAttributes(auth.AssertionAttributes())
+	if err != nil {
+		return nil, nil, ctxerr.Wrap(ctx, err, "invalid SSO attributes")
+	}
+
+	for _, teamRole := range ssoRolesInfo.Teams {
+		team, err := svc.ds.Team(ctx, teamRole.ID)
```

Comment:
Not sure sure if there is reason for concern - but maybe we should be careful about N+1 queries here?

Reply:
This is a check to perform on the readers before we attempt to do an invalid write (e.g. a team doesn't exist so a foreign key reference fails).

I agree it could be optimized, but maybe a-ok for now because we are not expecting customers to have hundreds of teams?

Am all ears if we really want to optimize this now.

Reply:
Yup sounds good to me, if `ssoRolesInfo.Teams` is small this shouldn't be a problem. Thanks